mautic/mautic
# Contact Token replacer

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Based on this discussion https://github.com/mautic/mautic/pull/6237#discussion_r224003824
This PR replace TokenHelper ::findLeadTokens static method to service easy to use for developers.

Contains:

- new abstract class TokenReplacer with Interface (https://github.com/mautic/mautic/pull/6710/files#diff-e13a9e67cbdbb48210a5eb6a1562ea44)
-  for contactfield and leadfield tokens we use new class ContactTokenReplacer extend TokenReplacer https://github.com/mautic/mautic/pull/6710/files#diff-e13a9e67cbdbb48210a5eb6a1562ea44
- support for localized date format by DateTimeHelper 

- use 3 new method
 - searchTokens - search tokens with raw attribute
 - getTokens - get tokens with replacet content
-  replaceTokens - replace all tokens from content

- contains test https://github.com/mautic/mautic/pull/6710/files#diff-f24664cac955f6e28e96e01147831f68

### DateTimeHelper localization support
- DateTime - doesn't support localizated months/days

> Docs: To format dates in other languages, you should use the setlocale() and strftime() functions instead of date() http://php.net/manual/en/function.date.php

But  strftime doesn't support date before Jan 1, 1970 on some OS

>  This means that %e, %T, %R and, %D (and possibly others) - as well as dates prior to Jan 1, 1970 - will not work on Windows, some Linux distributions, and a few other operating systems.  http://php.net/manual/en/function.strftime.php

I decide use DateTimeHelper as service with translations.  All DateTime classes loaded before works properly. Just If you want use getString (https://github.com/mautic/mautic/pull/6710/files#diff-36ed844b46ee77f22d7c9a23897b4053R196) You have to use DateTimeHelper as symfony service https://github.com/mautic/mautic/pull/6710/files#diff-b675f853d4cec144df96948057f29a34R88

### Another feature - relative and anniversary date modifier 
This require this PR https://github.com/mautic/mautic/pull/7228
This feature allow add generate relative or anniversary date to contact date

Usage

Relative date - will generate relative date to date stored in custom contact date field

- {contactfield=custom_date|datetime +7 days}
- {contactfield=custom_date|date +7 days} 

Anniversary date -  will generate anniversary date to date stored in custom contact date field. This is helpful If you want generate birthday coupon expired date (Hi, this is your birthday discount and it will expire in  {contactfield=custom_date|datetime anniversary +7 days})

- {contactfield=custom_date|date anniversary +7 days} 
- {contactfield=custom_date|datetime anniversary +7 days}


- tests included

#### Examples in screenshots  are for date 2018-01-02 

![image](https://user-images.githubusercontent.com/462477/52550668-52187b00-2dd9-11e9-8f61-20660ecb69a4.png)

![image](https://user-images.githubusercontent.com/462477/52550675-5e9cd380-2dd9-11e9-8e5b-f55a6d486ad9.png)





#### Steps to test this PR:
1.  Just test all channels use leadfield and contactfield tokens
- {leadfield=email}
- {contactfield=email}

Use also tokens with modifier
- {leadfield=email|true} - expect url encode result
- empty country field: {contactfield=country} - expect empty space
- empty country field: {contactfield=country|mydefaulttext} - expect mydefaultext
-  date field: {contactfield=mycustomdatefield|date} - expect formatted date
-  date field: {contactfield=mycustomdatefield|time} - expect formatted time
-  date field: {contactfield=mycustomdatefield|datetime} - expect formatted datetime
-  relative date field: {contactfield=mycustomdatefield|datetime +1 day} - expect formatted datetime + 1 day
-  relative date field: {contactfield=mycustomdatefield|datetime anniversary +1 day} - expect formatted datetime + 1 day with current year
-  try also load {contactfield=mycustomdatefield|date} on different localization of Mautic like english (french with translated messages.ini -should do it manually) with translated months/days (https://github.com/mautic/mautic/pull/6710/files#diff-e9975204439d1e820ce16d15658d4d48R554)

Channels:

Email
SMS
DynamicContent
Focus
Notification
Campaign action Send a Webhook 
Social - twitter send tweet
Page Redirect after form send



#### List deprecations along with the new alternative:
1. TokenHelper marked as deprecation and replace in all bundles and plugins with new service
